Content.Server/Ninja/Systems/SpaceNinjaSystem.cs

```diff
@@ -0,0 +1,358 @@
+    }
+
+    /// <summary>
+    /// Set up everything for ninja to work and send the greeting message/sound.
+    /// </summary>
+    /// <remarks>
+    /// Currently this adds the ninja to traitors, this should be removed when objectives are separated from traitors.
+    /// </remarks>
+    private void GreetNinja(Mind.Mind mind)
+    {
+        if (!_mind.TryGetSession(mind, out var session) || mind.OwnedEntity == null)
+            return;
+
+        // make sure to enable the traitor rule for the sweet greentext
+        var traitorRule = EntityQuery<TraitorRuleComponent>().FirstOrDefault();
+        if (traitorRule == null)
+        {
+            // TODO: fuck me this shit is awful, see TraitorRuleSystem
+            _gameTicker.StartGameRule("Traitor", out var ruleEntity);
+            traitorRule = Comp<TraitorRuleComponent>(ruleEntity);
+        }
+
+        var config = RuleConfig();
+        var role = new NinjaRole(mind, _proto.Index<AntagPrototype>("SpaceNinja"));
+        _mind.AddRole(mind, role);
+        // TODO: when objectives are not tied to traitor roles, remove this
+        _traitorRule.AddToTraitors(traitorRule, role);
+
+        // choose spider charge detonation point
+        // currently based on warp points, something better could be done (but would likely require mapping work)
+        var warps = new List<EntityUid>();
+        var query = EntityQueryEnumerator<WarpPointComponent, TransformComponent>();
+        var map = Transform(mind.OwnedEntity.Value).MapID;
+        while (query.MoveNext(out var warpUid, out var warp, out var xform))
+        {
+            // won't be asked to detonate the nuke disk or singularity or centcomm
+            if (warp.Location != null && !HasComp<PhysicsComponent>(warpUid) && xform.MapID == map)
+                warps.Add(warpUid);
+        }
+
+        if (warps.Count > 0)
+            role.SpiderChargeTarget = _random.Pick(warps);
+
+        // assign objectives - must happen after spider charge target so that the obj requirement works
+        foreach (var objective in config.Objectives)
+        {
+            AddObjective(mind, objective);
+        }
+
+        _audio.PlayGlobal(config.GreetingSound, Filter.Empty().AddPlayer(session), false, AudioParams.Default);
+        _chatMan.DispatchServerMessage(session, Loc.GetString("ninja-role-greeting"));
+    }
+
+    /// <summary>
+    /// Handle constant power drains from passive usage and cloak.
+    /// </summary>
+    private void UpdateNinja(EntityUid uid, SpaceNinjaComponent ninja, float frameTime)
+    {
+        if (ninja.Suit == null || !TryComp<NinjaSuitComponent>(ninja.Suit, out var suit))
+            return;
+
+        float wattage = _suit.SuitWattage(suit);
+
+        SetSuitPowerAlert(uid, ninja);
+        if (!TryUseCharge(uid, wattage * frameTime))
+        {
+            // ran out of power, reveal ninja
+            _suit.RevealNinja(ninja.Suit.Value, suit, uid);
+        }
+    }
```

Comment:
As above just make it generic.

I think we have PowerCellDraw for actively drawing power which also raises an event when done which may be better?